metrics: 1.0 update (#1360)
iterative/dvc.org: content/docs/command-reference/metrics/index.md

# metrics

A set of commands to add, manage, collect, and display project metrics:
[add](/doc/command-reference/metrics/add),
[show](/doc/command-reference/metrics/show),
[diff](/doc/command-reference/metrics/diff),
[modify](/doc/command-reference/metrics/modify), and
[remove](/doc/command-reference/metrics/remove).
A set of commands to manage, collect, and display _project metrics_:
[show](/doc/command-reference/metrics/show), and
[diff](/doc/command-reference/metrics/diff).

## Synopsis

```usage
usage: dvc metrics [-h] [-q | -v]

>                    {show,add,modify,remove,diff} ...

Reviewer: remove them from the highlighter

Reviewer 2: Already gone 🙂

Reviewer 2: It's config\prismjs\dvc-commands.js right?

Reviewer: yep, name sounds right

## Description

In order to track metrics associated to machine learning experiments, DVC has
the ability to mark a certain stage <abbr>outputs</abbr> as files containing
metrics to track.

> Metrics are project-specific floating-point or integer values e.g. AUC, ROC,

Reviewer: I think it's better for types of metrics go before Description in this case (and in plots)

Reviewer: at least the part that covers "we have two types ...."

Reviewer 2: I tried doing this and it didn't make much sense but I'm making some changes so Types comes earlier in plots and also use the term "scalar metrics" in metrics so it's obvious there's a distinction. See ee87e99f (sorry, also has a bunch of small updates to plots refs — related to this PR)

_false positives_, etc. They can be defined with the `-m` (`--metrics`) and `-M`

> options of `dvc run`.

Reviewer: include (--metrics-no-cache) or remote `--metrics`

Reviewer 2: Added ``(`--metrics-no-cache`)`` but Idk what you meant by "remote `--metrics`"

Reviewer: remote == remove

> ### Types of metrics

Reviewer: this should go before description

Reviewer 2: I didn't think you meant that literally earlier since it breaks our standard cmd ref structure. But OK, did it now in 3b898b37 (included also continuous and scalar terminology update discussed in Slack).

WDYT? I'm not so sure about this one...

Reviewer: > standard cmd ref structure.

we should not be following or trying to preserve it a cost of actual meaning and natural flow of the text. I don't know if this should go before or should be part of description, but introducing types and explaining the difference should be before explaining a particular type.

Reviewer 2: Would be tricky putting it in the Description before the basic explanation unless we remove the **Types of metrics** header and just make it into the first paragraph.

But for now I think it's fine actually, since this change only affects the index.md of both metrics and plots, which is kind of a placeholder cmd ref anyway. Other indices are used to explain concepts more than as cmd refs also, like for cache and remote 🙂 so we can be more flexible about these in terms of doc structure going fwd.

DVC has two concepts for metrics, that represent different results of machine
learning training or data processing:

1. `dvc metrics` represent **scalar numbers** such as AUC, _true positive rate_,
   etc.

> 2. `dvc plots` can be used to visualize **continuous metrics** such as AUC

Reviewer: continuous - @dmpetrov are we okay using the term continuous after all, even though it's not very mathematically correct?

Reviewer 2: FYI: The term "continuous metrics" is currently used 10 times.

   curves, loss functions, confusion matrices, etc.

In contrast to continuous metrics,
[scalar metrics](/doc/command-reference/metrics) should be stored in a
hierarchical files. Unlike its `dvc plot` counterpart, `dvc metrics diff` can
report the numeric difference between the metrics in different experiments, for
example an `AUC` metrics that is `0.801807` and gets increase by `+0.037826`:

```dvc
$ dvc metrics diff
    Path       Metric    Value      Change
summary.json   AUC      0.801807   0.037826
```

### Supported file formats

Metrics can be organized as tree hierarchies in JSON or YAML files. DVC
addresses specific metrics by the tree path. In the JSON example below, five
metrics are presented: `train.accuracy`, `train.loss`, `train.TN`, `train.FP`
and `time_real`.

```json
{
  "train": {
    "accuracy": 0.9886999726295471,
    "loss": 0.041855331510305405,
    "TN": 473,
    "FP": 845
  },
  "time_real": 344.61309599876404
}
```

DVC itself does not ascribe any specific meaning for these numbers. Usually they
are produced by the model training or model evaluation code and serve as a way
to compare and pick the best performing experiment.

> ### Default metrics files

Reviewer: metric files (metrics files sounds really bad from english lang perspective)

Reviewer: I would this into Description - it just a regular basic overview and it overlaps a lot with some parts in Description

Reviewer 2: I thought I got all of these before, seems like this one is missing. Yes. We have either "metric file", "metrics file" (which I'm not so sure of and may need reviewing, or "metric files" for plural.

Reviewer 2: Added this to #608

> ### Default metric files

Reviewer: should become part of the description

Reviewer: do we need a similar paragraph in plots?

Reviewer 2: It is part of the description.

Reviewer 2: > do we need a similar paragraph in plots?

No, plots require specifying a target/datafile

Reviewer 2: p.s. this led me to discover the following issues: #3925 and #3924

Reviewer 2: I mean iterative/dvc/issues/3925 and iterative/dvc/issues/3924

Reviewer: it overlaps a lot:

For example:

> Metric files can be added to `dvc.yaml` with the `--metrics` (`-m`) or
`--metrics-no-cache` (`-M`) options of `dvc run`, or manually to the `metrics`
section of a stage in `dvc.yaml`:

and

> This kind of metrics can be defined with the `-m` (`--metrics`) and `-M`
(`--metrics-no-cache`) options of `dvc run`.

Why can't we make it a single paragraph?

anyway, probably not a big deal .. we can keep it like this for now

Reviewer 2: > > do we need a similar paragraph in plots?
>
> No, plots require specifying a target/datafile

Actually probably we'll need it but this will be done as a result of resolving iterative/dvc/issues/3925 (I'll keep track).

> it overlaps a lot:

Yeah I didn't notice before. Removed redundancy in 47ad3116 (part of #1382)!

> Why can't we make it a single paragraph?

I could but would probably want to completely remove the dvc.yaml example in that case and the `cache:false` note. Also not sure it makes sense to have this info before the **Supported file formats** section. No strong opinion, I'll do it if you think it's best. Please lmk

In order to track metrics associated to machine learning experiments, DVC has
the ability to mark a certain stage <abbr>outputs</abbr> as files containing
metrics to track (see the `--metrics` option of `dvc run`). Metrics are

> project-specific floating-point or integer values e.g. `AUC`, `ROC`,
> `False-Positives` etc.

Reviewer 2: Probably no need for back quotes here. Maybe
```suggestion
project-specific floating-point or integer values e.g. AUC, ROC,
_false positives_, etc.
```

[Show](/doc/command-reference/metrics/show), and
[diff](/doc/command-reference/metrics/diff), commands are available to manage
<abbr>DVC project</abbr> metrics.

> 

Reviewer 2: Maybe remove these lines since they're already listed in the top definition.

> ## Formats and metrics types

Reviewer 2: ```suggestion
### Formats and types of metrics
```
(H3 if we agree this is part of the description so no need for extra entry in right sidebar)

> Supported file formats: JSON and YAML. Metrics can be organized in a tree
> hierarchy in a JSON file or a YAML file. DVC addresses the metrics by the tree
> path. In the example below five metrics are presented: `train.accuracy`,
> `train.loss`, `train.TN`, `train.FP` and `time_real`.

Reviewer 2: I always wanted to improve this paragrpah:
```suggestion
Metrics can be organized in a tree hierarchy in a JSON or YAML file. DVC
addresses specific metrics by the tree path. In the JSON example below,
five metrics are presented: `train.accuracy`, `train.loss`, `train.TN`, `train.FP`
and `time_real`.
```

Reviewer 2: Also applying this and other matching text updates to current dvc plot cmd ref...

```json
{
  "train": {
    "accuracy": 0.9886999726295471,
    "loss": 0.041855331510305405,
    "TN": 473,
    "FP": 845
  },
  "time_real": 344.61309599876404
}
```

DVC itself does not ascribe any specific meaning for these numbers. Usually
these numbers are produced by the model training or model evaluation code and
serve as a way to compare and pick the best performing experiment.

> ## Default metrics files

Reviewer 2: ```suggestion
### Default metrics files
```

> `dvc metrics show` and `dvc metrics diff` commands by default use all metrics
> files that are specified in `dvc.yaml`. So, there no need to specify metrics
> file name to see the deault metrics. The metrics files can be added to
> `dvc.yaml` by options `--metrics` (`-m`) or `--metrics-no-cache` (`-M`) of
> `dvc run` command or manualy to `metrics` section of a stage:

Reviewer 2: ```suggestion
`dvc metrics` subcommands use all metric files that are specified in
`dvc.yaml` by default. There's no need to specify metric file names to see
these metrics. Metric files can be added to `dvc.yaml` with the `--metrics`
(`-m`) or `--metrics-no-cache` (`-M`) options of `dvc run`, or manually to
the `metrics` section of a stage in `dvc.yaml`:
```

```yaml
stages:
  train:
    cmd: python train.py
    deps:
      - users.csv
    params:
      - epochs
      - dropout
      - lr
    outs:
      - model.pkl
    metrics:
      - summary.json:
          cache: false
```

> `False` value of the additional `cache` option specifies that this is not data
> file, it should not be moved to DVC cache and it should be commited to Git
> history instead.

Reviewer 2: ```suggestion
`cache: false` above specifies that `summary.json is not a data file: it will not
be <abbr>cached</abbr> by DVC. Metric files are normally committed with
Git instead.
```

### Difference between continuous and scalar metrics

> DVC has two concepts for metrics for representing result of machine learning
> training or data processing:

Reviewer 2: ```suggestion
DVC has two concepts for metrics, that represent different results of
machine learning training or data processing:
```

> 1. `dvc metrics` to represent scalar numbers such as AUC, true positive rate and
>    others.

Reviewer 2: ```suggestion
1. `dvc metrics` represent **scalar numbers** such as AUC, _true positive
rate_, etc.
```

> 2. `dvc plots` to visualize continuous metrics such as AUC curves, loss
>    functions, confusion matrices, and others.

Reviewer 2: ```suggestion
2. `dvc plots` can be used to visualize **continuous metrics** such as AUC curves,
loss functions, confusion matrices, etc.
```

> In this section, only scalar metrics are covered.

Reviewer 2: Probably not needed. I think it's obvious 🙂

## Examples

> First, let's create a simple DVC pipeline stage:

Reviewer 2: ```suggestion
First, let's create a simple stage:
```
Maybe link "stage" to the run cmd ref, if no other links to that doc exist nearby.

		}
```

> When metrics file changes in the user workspace witout commiting it to the Git,
> the `dvc metrics diff` command show the difference between metrics values:

Reviewer 2: It's "metric file(s)" right? The term we want, I mean. Singular
```suggestion
When there are metric file changes (before committing them with Git), the
`dvc metrics diff` command shows the difference between metrics values:
```

Reviewer 2: But maybe I can further simplify this paragraph to be less specific...

```dvc
$ dvc metrics diff
Path       Metric    Value    Change
eval.json  ACU       0.66729  0.01614
eval.json  error     0.16982  0.00322
eval.json  TP        516      -12
```

> Metrics commited to the Git history can be shown by referencing the commit name
> or any aliase:

Reviewer 2: ```suggestion
Metric files committed with Git can be shown by referencing the commit,
branch, or tag (or any other reference):
```
Link reference to the corresponding Git doc (revision I think? Need to check the other docs where we link to git-scm.com).

Reviewer 2: Or maybe skip the last part about refs.
